File: Week 18/java/1.DynamicProgramming.java
/*
 * WEEK 18 - DYNAMIC PROGRAMMING (Introduction)
 * Topic: DP Fundamentals — Memoization vs Tabulation
 *
 * DYNAMIC PROGRAMMING (DP):
 * Solve a problem by breaking it into overlapping subproblems,
 * solving each subproblem ONCE and storing the result.
 *
 * WHEN TO USE DP:
 * 1. Optimal substructure: optimal solution contains optimal solutions to subproblems.
 * 2. Overlapping subproblems: same subproblems are solved multiple times.
 *
 * TWO APPROACHES:
 * 1. TOP-DOWN (Memoization): Recursion + cache. Natural to think about.
 * 2. BOTTOM-UP (Tabulation): Fill a table from smallest subproblems upward. Usually faster.
 *
 * PROBLEMS COVERED:
 * 1. 0/1 Knapsack — classic DP on subsets
 * 2. Longest Common Subsequence (LCS)
 * 3. Coin Change (minimum coins to make amount)
 * 4. Subset Sum
 * 5. Climbing Stairs (Fibonacci-style DP)
 *
 * 0/1 KNAPSACK:
 * n items with weights[i] and values[i], capacity W.
 * Maximize total value without exceeding capacity.
 * Each item: either take it (0/1) or don't.
 *
 * Recurrence: dp[i][w] = max(dp[i-1][w], dp[i-1][w-weights[i]] + values[i])
 * Time: O(n*W), Space: O(n*W) or O(W) optimized
 */

import java.util.Arrays;

public class DynamicProgramming {

    // =====================
    // 0/1 KNAPSACK
    // =====================
    static int knapsack(int[] weights, int[] values, int W) {
        int n = weights.length;
        // dp[i][w] = max value using first i items with capacity w
        int[][] dp = new int[n + 1][W + 1];
        for (int i = 1; i <= n; i++) {
            for (int w = 0; w <= W; w++) {
                dp[i][w] = dp[i-1][w]; // don't take item i
                if (weights[i-1] <= w)  // take item i (if it fits)
                    dp[i][w] = Math.max(dp[i][w], dp[i-1][w - weights[i-1]] + values[i-1]);
            }
        }
        return dp[n][W];
    }

    // Space-optimized knapsack: O(W) space (process w in reverse!)
    static int knapsackOptimized(int[] weights, int[] values, int W) {
        int[] dp = new int[W + 1];
        for (int i = 0; i < weights.length; i++)
            for (int w = W; w >= weights[i]; w--) // REVERSE to avoid using item twice
                dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
        return dp[W];
    }

    // =====================
    // LONGEST COMMON SUBSEQUENCE (LCS)
    // =====================
    // Find length of longest subsequence present in both strings.
    // "ABCBDAB" and "BDCABA" → LCS length = 4 ("BCBA" or "BDAB")
    //
    // Recurrence:
    // dp[i][j] = dp[i-1][j-1] + 1              if s1[i-1] == s2[j-1]
    //          = max(dp[i-1][j], dp[i][j-1])    otherwise
    static int lcs(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++)
            for (int j = 1; j <= n; j++)
                if (s1.charAt(i-1) == s2.charAt(j-1))
                    dp[i][j] = dp[i-1][j-1] + 1;
                else
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
        return dp[m][n];
    }

    // =====================
    // COIN CHANGE (Minimum coins)
    // =====================
    // Minimum number of coins to make amount 'target'.
    // dp[a] = min coins to make amount a
    // dp[a] = min(dp[a - coin] + 1) for all coins where coin <= a
    static int coinChange(int[] coins, int target) {
        int[] dp = new int[target + 1];
        Arrays.fill(dp, target + 1); // initialize to "impossible"
        dp[0] = 0; // base case: 0 coins to make amount 0
        for (int a = 1; a <= target; a++)
            for (int coin : coins)
                if (coin <= a) dp[a] = Math.min(dp[a], dp[a - coin] + 1);
        return dp[target] > target ? -1 : dp[target];
    }

    // =====================
    // SUBSET SUM
    // =====================
    // Can we select a subset of 'nums' that sums to 'target'?
    static boolean subsetSum(int[] nums, int target) {
        boolean[] dp = new boolean[target + 1];
        dp[0] = true; // empty subset sums to 0
        for (int x : nums)
            for (int t = target; t >= x; t--) // reverse to avoid reusing
                dp[t] = dp[t] || dp[t - x];
        return dp[target];
    }

    // =====================
    // CLIMBING STAIRS
    // =====================
    // Reach stair n taking 1 or 2 steps at a time. How many distinct ways?
    static int climbStairs(int n) {
        if (n <= 2) return n;
        int prev2 = 1, prev1 = 2;
        for (int i = 3; i <= n; i++) {
            int curr = prev1 + prev2;
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }

    public static void main(String[] args) {
        // Knapsack
        System.out.println("=== 0/1 Knapsack ===");
        int[] weights = {2, 3, 4, 5};
        int[] values  = {3, 4, 5, 6};
        System.out.println("Max value (W=5):  " + knapsack(weights, values, 5));   // 7
        System.out.println("Max value (W=10): " + knapsack(weights, values, 10));  // 13
        System.out.println("Optimized (W=10): " + knapsackOptimized(weights, values, 10));

        // LCS
        System.out.println("\n=== LCS ===");
        System.out.println("LCS(\"ABCBDAB\",\"BDCABA\") = " + lcs("ABCBDAB","BDCABA")); // 4
        System.out.println("LCS(\"AGGTAB\",\"GXTXAYB\") = " + lcs("AGGTAB","GXTXAYB")); // 4

        // Coin Change
        System.out.println("\n=== Coin Change ===");
        System.out.println("coins=[1,2,5], amount=11: " + coinChange(new int[]{1,2,5}, 11)); // 3
        System.out.println("coins=[2], amount=3: " + coinChange(new int[]{2}, 3));            // -1
        System.out.println("coins=[1], amount=0: " + coinChange(new int[]{1}, 0));            // 0

        // Subset Sum
        System.out.println("\n=== Subset Sum ===");
        System.out.println("{3,1,5,9,12}, target=8:  " + subsetSum(new int[]{3,1,5,9,12}, 8));  // true
        System.out.println("{3,1,5,9,12}, target=7:  " + subsetSum(new int[]{3,1,5,9,12}, 7));  // false

        // Climbing Stairs
        System.out.println("\n=== Climbing Stairs ===");
        for (int i = 1; i <= 10; i++)
            System.out.println("climbStairs(" + i + ") = " + climbStairs(i));
        // 1,2,3,5,8,13,21,34,55,89 (Fibonacci!)
    }
}

File: Week 18/python/1.DynamicProgramming.py
"""
WEEK 18 - PYTHON DSA
Topic: Dynamic Programming Fundamentals
File: 1.DynamicProgramming.py

CONCEPT:
    Dynamic Programming solves a problem by breaking it into overlapping
    subproblems, solving each subproblem ONCE, and storing the answer.
    Two preconditions:
        1. Optimal substructure: optimal answer is built from optimal sub-answers.
        2. Overlapping subproblems: the same sub-call appears many times.

    Two implementation styles:
        TOP-DOWN  (memoization) - recursion + cache. Easy to derive from the
                                  natural recurrence; we use functools.cache.
        BOTTOM-UP (tabulation)  - fill a table from smallest to largest.
                                  Usually faster (no recursion / hash overhead);
                                  often allows space optimisations.

KEY POINTS:
    - Identify the STATE (parameters that uniquely identify a subproblem).
    - Write the RECURRENCE: dp[state] in terms of smaller states.
    - Establish the BASE CASE.
    - Decide ORDER OF EVALUATION (so dependencies are solved first).
    - Optimise SPACE if only the previous row(s) are needed.

ALGORITHM / APPROACH:
    Five canonical problems are covered:
        1. 0/1 Knapsack             (2-D DP, then 1-D space-optimised)
        2. Longest Common Subseq.   (2-D DP)
        3. Coin Change (min coins)  (unbounded coin DP)
        4. Subset Sum               (boolean DP)
        5. Climbing Stairs          (Fibonacci-style)

PYTHON-SPECIFIC NOTES:
    - `@functools.cache` (Py 3.9+) auto-memoises pure recursive functions.
    - List comprehensions like `[[0]*(W+1) for _ in range(n+1)]` build 2-D tables.
    - `math.inf` is a clean sentinel for "impossible".
    - Use `tuple` (not `list`) as cache keys when needed (lists aren't hashable).
    - Type hints from `typing` (List, Sequence, Optional) keep intent clear.

DRY RUN:
    Knapsack with weights=[2,3,4,5], values=[3,4,5,6], W=5
        dp table (rows = items 0..4, cols = capacities 0..5):
                 c=0 1 2 3 4 5
            i=0:   0 0 0 0 0 0
            i=1:   0 0 3 3 3 3        (item w=2,v=3)
            i=2:   0 0 3 4 4 7        (item w=3,v=4 ; dp[1][2]+4=7 at c=5)
            i=3:   0 0 3 4 5 7        (item w=4,v=5)
            i=4:   0 0 3 4 5 7        (item w=5,v=6 doesn't beat 7)
        Answer dp[4][5] = 7.

    LCS("ABCBDAB","BDCABA")
        Builds 7x6 table; final dp[7][6] = 4.
        Possible LCS strings: "BCBA", "BDAB".

    Coin change coins=[1,2,5], target=11:
        dp[0..11] starts with 0,inf,inf,...
        dp[1]=1, dp[2]=1, dp[3]=2, dp[4]=2, dp[5]=1, dp[6]=2,
        dp[7]=2, dp[8]=3, dp[9]=3, dp[10]=2, dp[11]=3.

COMPLEXITY:
    knapsack          O(n*W) time, O(n*W) space  (O(W) optimised)
    lcs               O(m*n) time, O(m*n) space
    coin_change       O(target * |coins|) time, O(target) space
    subset_sum        O(n * target) time, O(target) space
    climb_stairs      O(n) time, O(1) space
"""

from __future__ import annotations
import functools
import math
from typing import List, Sequence


# ---------------- 0/1 KNAPSACK ----------------
def knapsack(weights: Sequence[int], values: Sequence[int], W: int) -> int:
    """Bottom-up 2-D table; dp[i][w] = best value using first i items, capacity w."""
    n = len(weights)
    dp = [[0] * (W + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        wi, vi = weights[i - 1], values[i - 1]
        for w in range(W + 1):
            dp[i][w] = dp[i - 1][w]
            if wi <= w:
                dp[i][w] = max(dp[i][w], dp[i - 1][w - wi] + vi)
    return dp[n][W]


def knapsack_optimised(weights: Sequence[int], values: Sequence[int], W: int) -> int:
    """1-D rolling array. Iterate w in REVERSE to avoid using an item twice."""
    dp = [0] * (W + 1)
    for wi, vi in zip(weights, values):
        for w in range(W, wi - 1, -1):
            dp[w] = max(dp[w], dp[w - wi] + vi)
    return dp[W]


def knapsack_memo(weights: Sequence[int], values: Sequence[int], W: int) -> int:
    """Top-down memoised recursion."""
    n = len(weights)

    @functools.cache
    def best(i: int, cap: int) -> int:
        if i == n or cap == 0:
            return 0
        skip = best(i + 1, cap)
        take = best(i + 1, cap - weights[i]) + values[i] if weights[i] <= cap else 0
        return max(skip, take)

    return best(0, W)


# ---------------- LONGEST COMMON SUBSEQUENCE ----------------
def lcs(s1: str, s2: str) -> int:
    """Bottom-up tabulation."""
    m, n = len(s1), len(s2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    return dp[m][n]


def lcs_memo(s1: str, s2: str) -> int:
    """Top-down memoisation variant of LCS."""

    @functools.cache
    def go(i: int, j: int) -> int:
        if i == 0 or j == 0:
            return 0
        if s1[i - 1] == s2[j - 1]:
            return go(i - 1, j - 1) + 1
        return max(go(i - 1, j), go(i, j - 1))

    return go(len(s1), len(s2))


# ---------------- COIN CHANGE (min coins) ----------------
def coin_change(coins: Sequence[int], target: int) -> int:
    """Minimum coins to total `target`, or -1 if impossible."""
    INF = math.inf
    dp = [INF] * (target + 1)
    dp[0] = 0
    for a in range(1, target + 1):
        for c in coins:
            if c <= a and dp[a - c] + 1 < dp[a]:
                dp[a] = dp[a - c] + 1
    return -1 if dp[target] == INF else int(dp[target])


def coin_change_memo(coins: Sequence[int], target: int) -> int:
    @functools.cache
    def best(remaining: int) -> int:
        if remaining == 0:
            return 0
        if remaining < 0:
            return math.inf  # type: ignore[return-value]
        result = math.inf
        for c in coins:
            sub = best(remaining - c)
            if sub + 1 < result:
                result = sub + 1
        return result

    ans = best(target)
    return -1 if ans == math.inf else int(ans)


# ---------------- SUBSET SUM ----------------
def subset_sum(nums: Sequence[int], target: int) -> bool:
    """True if some subset sums exactly to `target`."""
    dp = [False] * (target + 1)
    dp[0] = True
    for x in nums:
        for t in range(target, x - 1, -1):
            if dp[t - x]:
                dp[t] = True
    return dp[target]


# ---------------- CLIMBING STAIRS ----------------
def climb_stairs(n: int) -> int:
    """Number of ways to climb n stairs taking 1 or 2 steps at a time."""
    if n <= 2:
        return n
    prev2, prev1 = 1, 2
    for _ in range(3, n + 1):
        prev2, prev1 = prev1, prev1 + prev2
    return prev1


def main() -> None:
    print("=== 0/1 Knapsack ===")
    weights, values = [2, 3, 4, 5], [3, 4, 5, 6]
    print("W=5  bottom-up:", knapsack(weights, values, 5))           # 7
    print("W=10 bottom-up:", knapsack(weights, values, 10))          # 13
    print("W=10 optimised:", knapsack_optimised(weights, values, 10))
    print("W=10 memoised :", knapsack_memo(weights, values, 10))

    print("\n=== LCS ===")
    print("LCS('ABCBDAB','BDCABA') =", lcs("ABCBDAB", "BDCABA"))     # 4
    print("LCS('AGGTAB','GXTXAYB') =", lcs("AGGTAB", "GXTXAYB"))     # 4
    print("LCS memo same input    =", lcs_memo("ABCBDAB", "BDCABA"))

    print("\n=== Coin Change ===")
    print("coins=[1,2,5], amount=11:", coin_change([1, 2, 5], 11))   # 3
    print("coins=[2], amount=3      :", coin_change([2], 3))         # -1
    print("coins=[1], amount=0      :", coin_change([1], 0))         # 0
    print("memoised, 11             :", coin_change_memo([1, 2, 5], 11))

    print("\n=== Subset Sum ===")
    print("[3,1,5,9,12] target=8:", subset_sum([3, 1, 5, 9, 12], 8))  # True
    print("[3,1,5,9,12] target=7:", subset_sum([3, 1, 5, 9, 12], 7))  # False

    print("\n=== Climbing Stairs ===")
    for i in range(1, 11):
        print(f"climb_stairs({i:2d}) = {climb_stairs(i)}")


if __name__ == "__main__":
    main()


"""
NOTES (Python vs Java):
    - Java's `int[][] dp = new int[n+1][W+1]` is one line; Python needs the
      list-comprehension `[[0]*(W+1) for _ in range(n+1)]` because
      `[[0]*(W+1)]*(n+1)` would alias the same inner list.
    - Memoisation in Java requires HashMap<Pair, Integer> or a big int[][]
      sentinel array; Python's `@functools.cache` is one decorator.
    - Use `math.inf` as a clean infinity sentinel; convert back via int().
    - The 1-D space-optimised knapsack iterates `w` in reverse for the SAME
      reason as Java: avoid double-counting the current item.
    - For huge state spaces, switch from @cache to manual dict-based memo
      to control eviction.
"""
